tests/Functional/TwoFactorAuthenticationTest.php

```diff
@@ -0,0 +1,254 @@
+    // Users that use a non-2FA authentication method and do not match the
+    // conditions, are free to log in.
+    $this->casLogin('basic_user@example.com', 'pwd1');
+    $this->assertUserLoggedIn();
+    $this->drupalLogout();
+    // Users that used a 2FA authentication method can always log in.
+    $this->casLogin('medium_user@example.com', 'pwd2');
+    $this->assertUserLoggedIn();
+    $this->drupalLogout();
+    $this->casLogin('high_user@example.com', 'pwd3');
+    $this->assertUserLoggedIn();
+    $this->drupalLogout();
+    $this->casLogin('medium_with_role_user@example.com', 'pwd6');
+    $this->assertUserLoggedIn();
+    $this->drupalLogout();
+
+    // Check that plugins are executed correctly, taking into account also
+    // the negate option.
+    $config->set('2fa_conditions', [
+      'oe_authentication_user_test' => [
+        'example' => TRUE,
+        'negate' => TRUE,
+        'id' => 'oe_authentication_user_test',
+      ],
+      'user_role' => [
+        'id' => 'user_role',
+        'negate' => FALSE,
+        'roles' => [
+          $role_one->id() => $role_one->id(),
+        ],
+      ],
+    ])->save();
+
+    $this->casLogin('role_one_user@example.com', 'pwd4');
+    $assert_session->statusMessageContains('You are required to log in using a two-factor authentication method.', 'error');
+    $this->casLogin('role_two_user@example.com', 'pwd5');
+    $this->assertUserLoggedIn();
+    $this->drupalLogout();
+    $this->casLogin('basic_user@example.com', 'pwd1');
+    $assert_session->statusMessageContains('You are required to log in using a two-factor authentication method.', 'error');
+  }
+
+  /**
+   * Tests that any exception thrown during condition evaluation is caught.
+   */
+  public function testConditionException(): void {
+    $config = \Drupal::configFactory()->getEditable('oe_authentication.settings');
+    $config
+      ->set('force_2fa', FALSE)
+      ->set('2fa_conditions', [
+        'oe_authentication_user_test' => [
+          'example' => TRUE,
+          'negate' => FALSE,
+          'id' => 'oe_authentication_user_test',
+        ],
+      ])
+      ->save();
+
+    \Drupal::state()->set('oe_authentication_user_test.crash_me', TRUE);
+    $basic_user = $this->createUser(name: 'basic_user');
+    $this->createCasUser('basic_user', 'basic_user@example.com', 'pwd1', [
+      'authenticationLevel' => 'BASIC',
+    ], $basic_user);
+    $this->casLogin('basic_user@example.com', 'pwd1');
+    $this->assertSession()->statusMessageContains('There was a problem validating your login, please contact a site administrator.', 'error');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function casLogin(string $email, string $password, array $query = []): void {
+    $this->drupalGet(Url::fromRoute('cas.login', [], ['query' => $query]));
+    $this->submitForm(['email' => $email, 'password' => $password], 'Login!');
+  }
+
+  /**
+   * Overrides the default logout method.
+   */
```

Comment:
Both of these method override the parent method, but one of them has `@inheritdoc`, the other has a custom doc description.
In Drupal world, I think all inherited methods have inheritdoc, but some have additional text.